azurerm_storage_share_file crashes with panic : makechan: size out of range

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

<!--- Please keep this note for the community --->

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

<!--- Thank you for keeping this note for the community --->


### Terraform Version

1.6.4

### AzureRM Provider Version

3.84.0

### Affected Resource(s)/Data Source(s)

azurerm_storage_share_file

### Terraform Configuration Files

```hcl
terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = ">= 3.70.0"
    }
  }
}

provider "azurerm" {
  features {}
}

locals {
  prefix = "my-prefix"
}


data "azurerm_virtual_network" "my_network" {
  name                = "my-network"
  resource_group_name = local.network_rg
}

locals {
  network_rg   = "my-network-rg"
  network_name = data.azurerm_virtual_network.my_network.name
  network_id   = data.azurerm_virtual_network.my_network.id

  location = "canadacentral"
}

resource "azurerm_log_analytics_workspace" "example" {
  name                = "${local.prefix}-law"
  location            = local.location
  resource_group_name = local.network_rg
  sku                 = "PerGB2018"
  retention_in_days   = 30
}

resource "azurerm_container_app_environment" "my_env" {
  name                       = "${local.prefix}-environment"
  location                   = local.location
  resource_group_name        = local.network_rg
  log_analytics_workspace_id = azurerm_log_analytics_workspace.example.id
}
resource "azurerm_container_app" "my_app" {
  name                         = "${local.prefix}-app"
  container_app_environment_id = azurerm_container_app_environment.my_env.id
  resource_group_name          = local.network_rg
  revision_mode                = "Single"

  template {
    container {
      name   = "nginx"
      image  = "index.docker.io/nginx:mainline-alpine3.18"
      cpu    = 0.25
      memory = "0.5Gi"

      volume_mounts {
        name = "mymount"
        path = "/usr/share/nginx/html"
      }
    }

    volume {
      storage_name = azurerm_storage_account.my_storage.name
      storage_type = "AzureFile"
      name         = "mymount"
    }
    container {
      name   = "examplecontainerapp"
      image  = "mcr.microsoft.com/azuredocs/containerapps-helloworld:latest"
      cpu    = 0.25
      memory = "0.5Gi"
    }
  }
}

resource "azurerm_storage_account" my_storage {
  name                     = replace("${local.prefix}-storage","-","")
  resource_group_name      = local.network_rg
  location                 = local.location
  account_tier             = "Standard"
  account_replication_type = "LRS"
}

resource "azurerm_storage_share" "example" {
  name                 = "sharename"
  storage_account_name = azurerm_storage_account.my_storage.name
  quota                = 5
}

resource "azurerm_storage_share_file" "nginx_index" {
  name             = "index.html"
  storage_share_id = azurerm_storage_share.example.id
  source           = "${path.module}/templates/index.html"
}

resource "azurerm_container_app_environment_storage" "my_fileshare" {
  name                         = "mycontainerappstorage"
  container_app_environment_id = azurerm_container_app_environment.my_env.id
  account_name                 = azurerm_storage_account.my_storage.name
  share_name                   = azurerm_storage_share.example.name
  access_key                   = azurerm_storage_account.my_storage.primary_access_key
  access_mode                  = "ReadOnly"
}
```


### Debug Output/Panic Output

```shell
╷
│ Error: Plugin did not respond
│
│   with module.my_container_app[0].azurerm_storage_share_file.nginx_index,
│   on ../../modules/my_container_app/main.tf line 80, in resource "azurerm_storage_share_file" "nginx_index":
│   80: resource "azurerm_storage_share_file" "nginx_index" {
│
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs
│ may contain more details.
╵

Stack trace from the terraform-provider-azurerm_v3.84.0_x5 plugin:

panic: makechan: size out of range

goroutine 223 [running]:
github.com/tombuildsstuff/giovanni/storage/2020-08-04/file/files.Client.PutFile({{{0xa58b7a0, 0xc00224c940}, {0xa58b760, 0xc00172c7b0}, 0x0, 0x0, 0x6fc23ac00, 0xd18c2e2800, 0x3, 0x6fc23ac00, ...}, ...}, ...)
        github.com/tombuildsstuff/giovanni@v0.20.0/storage/2020-08-04/file/files/range_put_file.go:37 +0x1a5
github.com/hashicorp/terraform-provider-azurerm/internal/services/storage.resourceStorageShareFileCreate(0x0?, {0x8e172c0?, 0xc0018a0000})
        github.com/hashicorp/terraform-provider-azurerm/internal/services/storage/storage_share_file_resource.go:186 +0x144a
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0xa5b8490?, {0xa5b8490?, 0xc002412ab0?}, 0xd?, {0x8e172c0?, 0xc0018a0000?})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/resource.go:766 +0x163
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc0015ab180, {0xa5b8490, 0xc002412ab0}, 0xc0021d89c0, 0xc001398a80, {0x8e172c0, 0xc0018a0000})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/resource.go:909 +0xa7e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc00030bb18, {0xa5b8490?, 0xc0024129c0?}, 0xc001e65090)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.29.0/helper/schema/grpc_provider.go:1060 +0xdbc
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc00191cb40, {0xa5b8490?, 0xc0024121b0?}, 0xc0021e4150)
        github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov5/tf5server/server.go:859 +0x56a
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x9921b00?, 0xc00191cb40}, {0xa5b8490, 0xc0024121b0}, 0xc0021e40e0, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.19.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:467 +0x169
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0003201e0, {0xa5c9ee0, 0xc001616b60}, 0xc0021f6480, 0xc0015b4cf0, 0x10161b78, 0x0)
        google.golang.org/grpc@v1.58.3/server.go:1374 +0xde7
google.golang.org/grpc.(*Server).handleStream(0xc0003201e0, {0xa5c9ee0, 0xc001616b60}, 0xc0021f6480, 0x0)
        google.golang.org/grpc@v1.58.3/server.go:1751 +0x9e7
google.golang.org/grpc.(*Server).serveStreams.func1.1()
        google.golang.org/grpc@v1.58.3/server.go:986 +0xbb
created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 55
        google.golang.org/grpc@v1.58.3/server.go:997 +0x145

Error: The terraform-provider-azurerm_v3.84.0_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```


### Expected Behaviour

no error

### Actual Behaviour

Was trying to setup a app container + app container environment with a mounted file share...

tried to do an apply and had the panic output (see above)

I suspect it maybe has something to do with the 3rs party library mentioned in output : 
`https://github.com/tombuildsstuff/giovanni/`

Note: this was initally run inside a module but I have removed it and placed it in the above as a single configuration file

### Steps to Reproduce

_No response_

### Important Factoids

_No response_

### References

_No response_

--- a/internal/services/storage/storage_share_file_resource.go
+++ b/internal/services/storage/storage_share_file_resource.go
@@ -175,6 +175,10 @@ func resourceStorageShareFileCreate(d *pluginsdk.ResourceData, meta interface{})
 			return fmt.Errorf("'stat'-ing File %q (File Share %q / Account %q): %+v", fileName, storageShareID.Name, storageShareID.AccountName, err)
 		}
 
+		if info.Size() == 0 {
+			return fmt.Errorf("file %q (File Share %q / Account %q) is empty", fileName, storageShareID.Name, storageShareID.AccountName)
+		}
+
 		input.ContentLength = info.Size()
 	}
 

--- a/internal/services/storage/storage_share_file_resource_test.go
+++ b/internal/services/storage/storage_share_file_resource_test.go
@@ -7,6 +7,7 @@ import (
 	"context"
 	"fmt"
 	"os"
+	"regexp"
 	"testing"
 
 	"github.com/hashicorp/terraform-provider-azurerm/internal/acceptance"
@@ -115,6 +116,23 @@ func TestAccAzureRMStorageShareFile_withFile(t *testing.T) {
 			),
 		},
 		data.ImportStep("source"),
+	})
+}
+
+func TestAccAzureRMStorageShareFile_withEmptyFile(t *testing.T) {
+	sourceBlob, err := os.CreateTemp("", "")
+	if err != nil {
+		t.Fatalf("Failed to create local source blob file")
+	}
+
+	data := acceptance.BuildTestData(t, "azurerm_storage_share_file", "test")
+	r := StorageShareFileResource{}
+
+	data.ResourceTest(t, r, []acceptance.TestStep{
+		{
+			Config:      r.withFile(data, sourceBlob.Name()),
+			ExpectError: regexp.MustCompile(`Error: file .* is empty`),
+		},
 	})
 }
 

--- a/website/docs/r/storage_share_file.html.markdown
+++ b/website/docs/r/storage_share_file.html.markdown
@@ -51,6 +51,8 @@ The following arguments are supported:
 
 * `source` - (Optional) An absolute path to a file on the local system. Changing this forces a new resource to be created.
 
+~> **Note** The file specified with `source` can not be empty.
+
 * `content_type` - (Optional) The content type of the share file. Defaults to `application/octet-stream`.
 
 * `content_md5` - (Optional) The MD5 sum of the file contents. Changing this forces a new resource to be created.

